hbase-client/src/main/java/org/apache/hadoop/hbase/security/provider/DefaultProviderSelector.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.security.provider;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.security.User;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@InterfaceAudience.Private
+public class DefaultProviderSelector implements AuthenticationProviderSelector {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultProviderSelector.class);
+  private final ReadWriteLock LOCK = new ReentrantReadWriteLock();
+
+  Configuration conf;
+  SimpleSaslClientAuthenticationProvider simpleAuth = null;
+  GssSaslClientAuthenticationProvider krbAuth = null;
+  DigestSaslClientAuthenticationProvider digestAuth = null;
+
+  @Override
+  public void configure(Configuration conf, Map<Byte,SaslClientAuthenticationProvider> providers) {
+    try {
+      LOCK.writeLock().lock();
```

Comment:
Will there be contention while calling this method? and that's why we need LOCK here?

Reply:
Can `final Configuration conf` work here? Not very clear to me here.

Reply:
I added this on the last set of reviews to work around findbugs complaints. We can't mark `conf` as `final` here since this isn't in the constructor.

It's possible that I could add a findbugs exception since we have expectations on how the callers of this object work. I'm not sure if this is a "hot enough" path to cause performance issues (it should only be called when setting up a new connection to a RegionServer/Master, not for every RPC executed).

Related: I should ear-mark some time to run a basic `hbase pe` to see if any big regression is introduced.